internal/container/container.go

```diff
@@ -0,0 +1,249 @@
+	}
+
+	return nil
+}
+
+// Start starts an existing container with the given name.
+func Start(ctx context.Context, name string) error {
+	cli, err := newClient()
+	if err != nil {
+		return err
+	}
+
+	err = cli.ContainerStart(ctx, name, container.StartOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to start container: %w", err)
+	}
+
+	return nil
+}
+
+// Exec runs a command in the given container, stream to stderr. Return an error if the command fails.
+func Exec(ctx context.Context, name string, command ...string) error {
+	logger := log.FromContext(ctx)
+	logger.Debugf("Running command %q in container %q", command, name)
+
+	cli, err := newClient()
+	if err != nil {
+		return err
+	}
+
+	exec, err := cli.ContainerExecCreate(ctx, name, types.ExecConfig{
+		Cmd:          command,
+		AttachStderr: true,
+		AttachStdout: true,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create exec: %w", err)
+	}
+
+	attach, err := cli.ContainerExecAttach(ctx, exec.ID, types.ExecStartCheck{})
+	if err != nil {
+		return fmt.Errorf("failed to attach exec: %w", err)
+	}
+	defer attach.Close()
+
+	_, err = io.Copy(os.Stderr, attach.Reader)
+	if err != nil {
+		return fmt.Errorf("failed to stream exec: %w", err)
+	}
+
+	err = cli.ContainerExecStart(ctx, exec.ID, types.ExecStartCheck{})
+	if err != nil {
+		return fmt.Errorf("failed to start exec: %w", err)
+	}
+
+	inspect, err := cli.ContainerExecInspect(ctx, exec.ID)
+	if err != nil {
+		return fmt.Errorf("failed to inspect exec: %w", err)
+	}
+	if inspect.ExitCode != 0 {
+		return fmt.Errorf("exec failed with exit code %d", inspect.ExitCode)
+	}
+
+	return nil
+}
+
+// GetContainerPort returns the host TCP port of the given container's exposed port.
+func GetContainerPort(ctx context.Context, name string, port int) (int, error) {
+	cli, err := newClient()
+	if err != nil {
+		return 0, err
+	}
+
+	inspect, err := cli.ContainerInspect(ctx, name)
+	if err != nil {
+		return 0, fmt.Errorf("failed to inspect container: %w", err)
+	}
+
+	containerPort := fmt.Sprintf("%d/tcp", port)
+	hostPort, ok := inspect.NetworkSettings.Ports[nat.Port(containerPort)]
+	if !ok {
+		return 0, fmt.Errorf("container port %q not found", containerPort)
+	}
+
+	if len(hostPort) == 0 {
+		return 0, fmt.Errorf("container port %q not bound", containerPort)
+	}
+
+	return nat.Port(hostPort[0].HostPort).Int(), nil
+}
+
+// PollContainerHealth polls the given container until it is healthy or the timeout is reached.
+func PollContainerHealth(ctx context.Context, containerName string, timeout time.Duration) error {
+	logger := log.FromContext(ctx)
+	logger.Debugf("Waiting for %s to be healthy", containerName)
+
+	cli, err := newClient()
+	if err != nil {
+		return err
+	}
+
+	pollCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	for {
+		select {
+		case <-pollCtx.Done():
+			return fmt.Errorf("timed out waiting for container to be healthy: %w", pollCtx.Err())
+
+		case <-time.After(1 * time.Millisecond):
```

Comment:
Maybe make this slightly less enthusiastic, say 100ms?

Reply:
I did wonder why this was so keen.